src/impl/apis/data_api_impl.py

```diff
@@ -313,135 +435,59 @@ async def run_query(
+    base_table_name, table_name = await get_table_name(
+        query_parameters, pool=pool, token_model=token_model
+    )
+    tr = await get_temporal_resolution(query_parameters.trid, token_model=token_model, pool=pool)
+    metadata = await get_metadata(query_parameters, tr, pool, token_model)
+    query_stream_generator = stream_query(base_table_name, metadata, pool, table_name)
     if query_parameters.category_id.lower() in ["residents", "presence"]:
-        out = region_to_csv(result.data_by_date)
+        async for csv_chunk in stream_region_to_csv(query_stream_generator):
+            yield csv_chunk
     elif query_parameters.category_id.lower() in ["relocations", "movements"]:
-        out = flows_to_csv(result.data_by_date)
+        async for csv_chunk in stream_flows_to_csv(query_stream_generator):
+            yield csv_chunk
     else:
         raise HTTPException(
             status_code=HTTPStatus.NOT_IMPLEMENTED,
             detail=f"CSV not yet implemented for {query_parameters.category_id.lower()}",
         )
-    return out
-
-
-def region_to_csv(region_data: Dict[str, Dict[str, float]]) -> str:
-    csv_rows = []
-    for date, data in region_data.items():
-        for source_region, value in data.items():
-            row = {"date": date, "area_code": source_region, "value": value}
-            csv_rows.append(",".join(str(v) for v in row.values()))
-    csv_header = ",".join(row.keys())
-    csv_out = "\r\n".join((csv_header, *csv_rows))
-    return csv_out
-
-
-def flows_to_csv(flow_data: Dict[str, Dict[str, Dict[str, float]]]) -> str:
-    csv_rows = []
-    for date, data in flow_data.items():
-        for source_region, value in data.items():
-            for dest_region, flow_value in value.items():
-                row = {
-                    "date": date,
-                    "origin_code": source_region,
-                    "destination_code": dest_region,
-                    "value": flow_value,
-                }
-                csv_rows.append(",".join(str(v) for v in row.values()))
-    csv_header = ",".join(row.keys())
-    csv_out = "\r\n".join((csv_header, *csv_rows))
-    return csv_out
+
+
+async def stream_region_to_csv(
+    region_stream: AsyncGenerator,
+) -> AsyncGenerator[str, None]:
+    yield "date,area_code,value\r\n"
+    async for chunk in region_stream:
+        if chunk:
+            yield "\r\n".join(
+                f"{row['dt'].strftime('%Y-%m-%d')},{row['spatial_unit_id']},{row['data']}"
+                for row in chunk
+            ) + "\r\n"
+    yield "\r\n"
+
+
+async def stream_flows_to_csv(flow_stream: AsyncGenerator) -> AsyncGenerator[str, None]:
+    yield "date,origin_code,destination_code,value\r\n"
+    async for chunk in flow_stream:
+        if chunk:
+            yield "\r\n".join(
+                f"{row['dt'].strftime('%Y-%m-%d')},{row['origin']},{row['destination']},{row['data']}"
+                for row in chunk
+            ) + "\r\n"
+    yield "\r\n"
```

Comment:
Do we need the final newline?

Reply:
Good practice for CSVs, I thought?
(that famously well-specified format)

Reply:
On thinking, most of this is going to be consumed by various Pandas or Excels, so I'll drop it.